### PR TITLE
[rush] Rush should return a nonzero exit code if warnings are encountered during a build

### PR DESCRIPTION
**Pull request by iclanton**

### CODE REVIEW

**Comment by octogonz** [comment]
The behavior you're introducing is the intended behavior of Rush, which was designed that way specifically to avoid warnings creeping in to the master build by accident.  The [Rush docs](https://rushjs.io/pages/maintainer/add_to_repo/) even describe this behavior:

> There are a few things to keep in mind when creating a `"build"` script:
>
> . . .
>
> * If the process returns a non-zero exit status, Rush will assume there was a failure and will block downstream builds.
>
> * If the command writes anything to the `stderr` stream, Rush will interpret this to mean that at least one error or warning was reported.  This will break the build.  (This is by design -- if you allow people to merge PRs that "cry wolf", pretty soon you will find that so many warnings have accumulated that nobody even reads them any more.)  Some tooling libraries (e.g. Jest) write to `stderr` as part of their normal operation; you will need to [redirect their output](https://github.com/Microsoft/web-build-tools/blob/master/core-build/gulp-core-build/src/tasks/JestReporter.ts).
>

Thus the issue you're fixing is actually a regression (that apparently was overlooked for a fairly long time).  Since some people may have come to rely on this behavior, I'd propose to fix it like this:

- Restore the intended behavior, where the Rush process returns a nonzero exit code if any of the project builds wrote anything to stderr.  (Note that this doesn't block downstream projects from building, since we want to fail a CI job without interfering with local developer builds.)

- Introduce a **rush.json** setting that allows people to "opt out" of this behavior, if their toolchain came to rely on the current behavior of Rush.  (e.g. `allowWarningsInSuccessfulBuild`) The doc comment for the setting should explain that opting out is discouraged.

- Release this fix as a minor version bump for Rush, with a conspicuous changelog entry (e.g. `(BEHAVIOR CHANGE) Fix a regression that enabled CI jobs to succeed even if a task wrote warnings to stderr; see "allowWarningsInSuccessfulBuild" in rushj.son for details`)

**Comment by octogonz** [comment]
Is `allowWarningsInSuccessfulBuild` relevant to custom commands with `commandKind=global`?

File: apps/rush-lib/assets/rush-init/common/config/rush/command-line.json [hunk 1]
```diff
@@ -78,7 +78,29 @@
...
+
+      /**
+       * When invoking shell scripts, Rush uses a heuristic to distinguish errors from warnings:
+       * - If the shell script returns a nonzero process exit code, Rush interprets this as "one or more errors".
+       * Error output is displayed in red, and it prevents Rush from attempting to process any downstream projects.
+       * - If the shell script returns a zero process exit code but writes something to its stderr stream,
+       * Rush interprets this as "one or more warnings". Warning output is printed in yellow, but does NOT prevent
+       * Rush from processing downstream projects.
+       *
+       * Thus, warnings thus do not interfere with local development, but they will cause a CI job to fail, because
```
Comment (octogonz) [inline apps/rush-lib/assets/rush-init/common/config/rush/command-line.json]: Fix "Thus, warnings thus"

File: apps/rush-lib/src/cli/RushCommandLineParser.ts [hunk 2]
```diff
@@ -247,7 +249,8 @@ export class RushCommandLineParser extends CommandLineParser {
 
             enableParallelism: command.enableParallelism,
             ignoreMissingScript: command.ignoreMissingScript || false,
-            ignoreDependencyOrder: command.ignoreDependencyOrder || false
+            ignoreDependencyOrder: command.ignoreDependencyOrder || false,
+            doNotFailOnWarnings: !!command.allowWarningsInSuccessfulBuild
```
Comment (octogonz) [inline apps/rush-lib/src/cli/RushCommandLineParser.ts]: Is there any difference between `doNotFailOnWarnings` vs `allowWarningsInSuccessfulBuild`?  If not, let's use a consistent name everywhere

**Comment by iclanton** [comment]
This change fixes https://github.com/microsoft/web-build-tools/issues/1329

**Pull request merged**